lgc/builder/BuilderImpl.cpp

```diff
@@ -329,6 +330,114 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+    const SmallVector<Instruction *, 64> &indexToInstr;
+
+  public:
+    iterator(BitVector::const_set_bits_iterator it, const IndexToInstructionVec &indexToInstr)
+        : it(it), indexToInstr(indexToInstr) {}
+    iterator &operator++() {
+      ++it;
+      return *this;
+    }
+
+    Instruction *operator*() {
+      unsigned index = *it;
+      assert(index < indexToInstr.size() && "Index out of range.");
+      return indexToInstr[index];
+    }
+
+    bool operator!=(const iterator &otherIt) {
+      assert(&otherIt.indexToInstr == &indexToInstr && "Iterators of different objects.");
+      return otherIt.it != it;
+    }
+  };
+
+  iterator begin(const IndexToInstructionVec &indexToInstr) const {
+    return iterator(bits.set_bits_begin(), indexToInstr);
+  }
+
+  iterator end(const IndexToInstructionVec &indexToInstr) const { return iterator(bits.set_bits_end(), indexToInstr); }
+
+  void insert(Instruction *instr, const InstrToIndexMap &instrToIndex) {
+    auto it = instrToIndex.find(instr);
+    assert(it != instrToIndex.end() && "Expected to find instr in instrToIndex.");
+    unsigned index = it->second;
+    bits.set(index);
+  }
+
+  unsigned size() const { return bits.size(); }
+
+  bool empty() const { return bits.empty(); }
+};
+
+class TraceNonUniformIndex {
+  // Maps the instruction to its index in the bit vector.
+  TinyInstructionSet::InstrToIndexMap instrToIndex;
+  // The instructions used as keys in instrToIndex in program order. It is used to map an index to an instruction.
+  TinyInstructionSet::IndexToInstructionVec indexToInstr;
+  // Maps an instruction to its dependencies.
+  DenseMap<Value *, TinyInstructionSet> instrDeps;
+  bool scalarizeDescriptorLoads;
+  unsigned upperLimit;
+  void insertNewValueInInstrDeps(Value *, Instruction *);
+
+public:
+  TraceNonUniformIndex(Instruction *nonUniformInst, bool scalarizeDescriptorLoads = false, unsigned upperLimit = 64)
+      : scalarizeDescriptorLoads(scalarizeDescriptorLoads), upperLimit(upperLimit) {
+    // Initialization of instrToIndex and indexToInstr.
+    if (scalarizeDescriptorLoads) {
+      unsigned cnt = 0;
+      for (Instruction *I = nonUniformInst->getPrevNode(); I != nullptr && cnt < upperLimit;
+           I = I->getPrevNode(), ++cnt) {
+        indexToInstr.push_back(I);
+        instrToIndex[I] = cnt;
+      }
+    }
+  }
+
+  Value *run(Value *);
+
+  const DenseMap<Value *, TinyInstructionSet> &getInstrDeps() const { return instrDeps; }
+
+  const TinyInstructionSet::IndexToInstructionVec &getIndexToInstr() const { return indexToInstr; }
+
+  bool foundDependencies() const { return scalarizeDescriptorLoads; }
+};
+
+void TraceNonUniformIndex::insertNewValueInInstrDeps(Value *newValue, Instruction *currentVisitedInstr) {
+  if (!instrToIndex.contains(currentVisitedInstr)) {
+    // The instruction is either outside of 64 limit or in a different basic block. So, we bail-out scalarization.
+    scalarizeDescriptorLoads = false;
```

Comment:
It seems setting TraceNonUniformIndex.scalarizeDescriptorLoads to `false` here means that the next time TraceNonUniformIndex::run is invoked it will not try to scalarize. That means if there are two non-uniform operands of, say, image.sample, if the first one could not be scalarized, then the second one would not be either.

Reply:
This is how the existing code works. I am working to change it.

Reply:
Fair enough. Can you extend the comment to make it clear that this will also affect scalarization of other operands?

Implementation-wise I guess it is fine to address it in a follow-up commit to keep the scope of the changes smaller.

Reply:
Done in https://github.com/GPUOpen-Drivers/llpc/pull/2859